### PR TITLE
make: disable dwarf compression on CI builds

### PR DESCRIPTION
**Pull request by kaspar030**
### Contribution description

If enabled, ccache on CI becomes almost ineffective.

This PR guards the CFLAG with RIOT_CI_BUILD.

From #12110 

```
This somehow makes ccache almost ineffective on CI. Maybe you noticed almost doubled build times today? My first thought turned out to be wrong. During the day, I cleared the cache on one of the workers, but build times stayed double. So cache thrashing was no issue, (and should't be actually as even the oldest PRs get rebased to master and are thus not much different).

Then I ran a script that ran a limited build (samr21-xpro only) on one worker, once with clean cache, then one more time with hot cache, on all merged commits since the last nightly.

Long story short, 3aa8bc0 (this PR's merge commit) somehow makes ccache ineffective.
```

I tried to reproduce locally with this:

```
cd examples/gnrc_networking; hyperfine -w1 "RIOT_CI_BUILD=1 BOARD=samr21-xpro make clean all -j8"
```

On my laptop, master takes ~2.6s, this PR ~2.2s, so the difference is not so drastic. Let's see how quickly this PR builds on CI. Might need two builds (cold vs hot cache) to see the difference.

### Issues/PRs references

#12150

### CODE REVIEW

File: makefiles/cflags.inc.mk [hunk 1]
```diff
@@ -39,9 +39,10 @@ endif
 CFLAGS += -fno-common
 
 # Compress debug info. This saves approximately 50% of disk usage.
-# It has no effect if debugging information is not emitted, so it can be left
-# on unconditionally.
-OPTIONAL_CFLAGS += -gz
+# This makes ccache less effective, so disable for CI builds.
+ifneq (1, $(RIOT_CI_BUILD))
```
Comment (miri64) [inline makefiles/cflags.inc.mk]: If `ccache` is the problem and not CI, then the guard should be `ccache` specific, no?
```suggestion
ifeq (, $(CCACHE))
```

Comment (kaspar030) [reply makefiles/cflags.inc.mk]: That seems reasonable, but on my laptop, the performance hit is only ~20%, not 50%. It is quite beefy compared to the multicore CI workers (per-core wise). This might be more affecting build performance on those CPU or memory bound setups.

Also, ccache doesn't compress cache contents by default ("compression=false"), so this PR does save 50% space. (even when no debug symbols are generated...) CI enables ccache compression, which already brings down cache content size to the same level as with "-gz".

So I think locally this is might be useful.

That said, shouldn't using ccache be the first hint to every RIOT developer? Is any of you guys *waiting for local single builds* longer than a couple of seconds??

Comment (miri64) [reply makefiles/cflags.inc.mk]: Mh... Then maybe a combination of the two. I did not consider that `ccache` is also run locally sometimes (no, I don't use it, only when doing multi-board builds, because having just a small-ish SSD in my PC here, I value HDD space over time (also I have an octo-core so it is pretty fast anyway) ;-P). I was just thinking that the dependency is not "the CI", but `ccache` in this instance. There are other CIs used in our ecosystem than Murdock and not all use `ccache`. 

Maybe a combination of the two then?

```suggestion
ifneq (1,$(RIOT_CI_BUILD))
ifeq (,$(CCACHE))
```

Comment (kaspar030) [reply makefiles/cflags.inc.mk]: > having just a small-ish SSD in my PC here, I value HDD space over time (also I have an octo-core so it is pretty fast anyway) ;-P).

Does it have a lot of RAM? 4-8G of ccache in tmpfs makes that octocore FLY!

Comment (miri64) [reply makefiles/cflags.inc.mk]: Good thinking!

**Comment by jcarrano** [comment]
I'm OK with this change, but let's make it clear it is a hack, since we do not know the root cause of the slowdown (or do we?).

At first sight it does not make sense, especially if debugging info is not being generated (then the -gz is a no-op). Should we report a bug to ccache?

File: makefiles/cflags.inc.mk (hunk 1 above)
Comment (kaspar030) [reply makefiles/cflags.inc.mk]: I'm in the process of nailing down the problem within the ccache source.
It seems like as soon as any "-g*" option is used, ccache assumes debug info is generated, and includes the current cwd in the hash that identifies cache entries. For "-gz" without any other "-g" other than "-g0", this is a ccache bug. CI builds with a seperate builddir per checkout, so that could be the reason why it is affected a lot more than a local build.

I'm just now trying to create a reproducer and maybe open an upstream bug. For now, I'm confident this is a corner case of how ccache is used in our CI...

Hm, I'm realizing that local builds in different build directories would be affected too. I do that all the time. :) So maybe guarding that by ccache is a good idea.

**Comment by kaspar030** [comment]
> At first sight it does not make sense, especially if debugging info is not being generated (then the -gz is a no-op). Should we report a bug to ccache?

I'm struggling to create a reproducer, but the ccache bug issue page is already open.

**Comment by kaspar030** [comment]
https://github.com/ccache/ccache/issues/464

**Comment by jcarrano** [comment]
> ccache/ccache#464

Cool. Let's add all this info to the commit message and perhaps a comment in the makefile. Otherwise we will lose track of this issue. With that I believe we can merge this.

**Comment by kaspar030** [comment]
So the gist of this is:

1. if *any* debug option starting with "-g" is used, ccache adds the current working directory to the compilation hash, unless base_dir is set and matches cwd *and* ```-fdebug-prefix-map``` is used.

2. RIOT does set the basedir correctly, but not configure ```-fdebug-prefix-map```

3. without "-gz", ccache correctly strips the basedir from the absolute path's that RIOT's build system uses and thus caches between builds on CI

4. with "-gz", ccache assumes it is generating debug info and includes the cwd hash, breaking cross-build caching

**Comment by kaspar030** [comment]
> Let's add all this info to the commit message and perhaps a comment in the makefile. Otherwise we will lose track of this issue.

Done. Let's discuss the space saving vs. local cross-directory ccache effectiveness another time.

File: makefiles/cflags.inc.mk (hunk 1 above)
Comment (jcarrano) [reply makefiles/cflags.inc.mk]: I second @miri64 , the test should not be for RIOT_CI_BUILD. The current meaning of RIOT_CI_BUILD is "turn off any non-determinism". I may want to build with both RIOT_CI_BUILD and debug. Also, the issue may not manifest itself if the path is not changed, which means even testing for CCACHE is too much.

This should be done in a way that is specific to murdock and nothing else.

**Comment by cladmi** [comment]
Please do not put it as `RIOT_CI_BUILD` triggered. I want compression on other CI builds. Here it is a `murdock` issue.
The error is caused by having the `CFLAGS_DBG` overwritten to `""` in `.murdock` so put the fix there.

**Comment by kaspar030** [comment]
> The error is caused by having the `CFLAGS_DBG` overwritten to `""` in `.murdock` so put the fix there.

How about not adding "-gz" when not even generating debug output? IMO, "-gz" belongs into "CFLAGS_DBG".

murdock setting ```CFLAGS_DBG=""``` is it's method of disabling a debug build. Suddenly, globally, setting "-gz" *breaks stuff*, as we have seen, so it should not been done in the first place.

I suggest merging this workaround (getting back the behaviour of yestarday), and anyone who wants "-gz" comes up with a clean solution.

**Comment by kaspar030** [comment]
> The current meaning of RIOT_CI_BUILD is "turn off any non-determinism".

That's wrong, RIOT_CI_BUILD actually *enables* deterministic builds.

**Comment by miri64** [comment]
> > The current meaning of RIOT_CI_BUILD is "turn off any non-determinism".
> 
> That's wrong, RIOT_CI_BUILD actually _enables_ deterministic builds.

You just sait the same ;-) Just negated.

**Comment by kaspar030** [comment]
> in `.murdock` so put the fix there.

Do you have a suggestion on how to do that? I think ```OPTIONAL_CFLAGS_BLACKLIST="-gz" make ..." will fix that variable to one value.

**Comment by jcarrano** [comment]
> That's wrong, RIOT_CI_BUILD actually enables deterministic builds.

Yes, that's what I said. -gz has nothing to do with deterministic builds.

> IMO, "-gz" belongs into "CFLAGS_DBG".

The reason I added it to CFLAGS_OPTIONAL is to be able to filter it out with OPTIONAL_CFLAGS_BLACKLIST, and because I enabling compression is orthogonal to the other debug settings.

You can take out -gz by adding it to OPTIONAL_CFLAGS_BLACKLIST.

**Comment by jcarrano** [comment]
`OPTIONAL_CFLAGS_BLACKLIST="-gz"` should work fine. All assignments to this are made via `+=`.

**Comment by kaspar030** [comment]
> `OPTIONAL_CFLAGS_BLACKLIST="-gz"` should work fine. All assignments to this are made via `+=`.

```
$ OPTIONAL_CFLAGS_BLACKLIST="-gz" BOARD=arduino-mega2560 make info-debug-variable-OPTIONAL_CFLAGS
-fno-delete-null-pointer-checks -fdiagnostics-color -Wstrict-prototypes -Wold-style-definition -gz -Wformat=2 -Wformat-overflow -Wformat-truncation
```

**Comment by kaspar030** [comment]
Seriously, "-gz" *only* makes sense when used together with any other "-g" othe than "-g0". Let's move it to "CFLAGS_DBG".

**Comment by jcarrano** [comment]
> `$ OPTIONAL_CFLAGS_BLACKLIST="-gz" BOARD=....`

yeah, thats OK. the relevant piece of code is this:

```
CFLAGS += $(filter-out $(OPTIONAL_CFLAGS_BLACKLIST),$(OPTIONAL_CFLAGS))
```

So stuff does not get removed from OPTIONAL_CFLAGS, it only gets filtered out when it is added to CFLAGS.

**Comment by kaspar030** [comment]
I still suggest merging this as is and revert/fix properly later. PR's are queuing up in Murdock, everyone is waiting.

**Comment by miri64** [comment]
> I still suggest merging this as is and revert/fix properly later. PR's are queuing up in Murdock, everyone is waiting.

At least the suggested dependency fix should be applied ;-).

**Comment by miri64** [comment]
@kaspar030 please have a look at #12202. This seems to me the more cleaner hack.

**Comment by kaspar030** [comment]
I'm ok with #12202. 

Can we still get consensus on whether "-gz" should be default? It does lower compilation speed by ~20% on my laptop.

**Comment by kaspar030** [comment]
> ccache/ccache#464

FIY, the upstream bug has already been fixed (in ccache 3.7.4 and upwards).
